### PR TITLE
feat: paginate push reconcile

### PR DESCRIPTION
**Pull request by adityathebe**

### CODE REVIEW

File: db/upstream.go [hunk 1]
```diff
@@ -2,69 +2,86 @@
...
+		)
+		SELECT
+			encode(digest(string_agg(id::TEXT, ''), 'sha256'), 'hex') as sha256sum,
+			MAX(id) as last_id,
+			COUNT(*) as total
+		FROM
+			id_list`, table)
+
+	var resp api.PaginateResponse
+	err := Gorm.WithContext(ctx).Raw(query, from, size).Scan(&resp).Error
```
[the remark this answers is not in the document]
Comment (adityathebe) [reply db/upstream.go:39]: This check `collections.Contains(api.TablesToReconcile, table)` ensures that the input is controlled

File: db/upstream.go [hunk 2]
```diff
@@ -5,66 +5,72 @@
...
-	var configItems []models.ConfigItem
-	if err := Gorm.Select("id").Where("agent_id = ?", agentID).Find(&configItems).Pluck("id", &response.ConfigItems).Error; err != nil {
-		return nil, err
-	}
+func GetAllResourceIDsOfAgent(ctx context.Context, agentID string, req api.PushPaginateRequest) ([]string, error) {
+	var response []string
 
-	return &response, nil
+	query := fmt.Sprintf("SELECT id FROM %s WHERE agent_id = ? AND id > ? ORDER BY id LIMIT ?", req.Table)
+	err := Gorm.WithContext(ctx).Raw(query, agentID, req.From, req.Size).Scan(&response).Error
```
[the remark this answers is not in the document]
Comment (adityathebe) [reply db/upstream.go:83]: This check `collections.Contains(api.TablesToReconcile, table)` ensures that the input is controlled

**Comment by moshloop** [comment]
@adityathebe are there tests that cover this? We might want to make the page size configurable as well

**Comment by yashmehrotra** [comment]
@adityathebe Can we also handle resource deletions here as well ?

Something like, if it is present in upstream, but not in the agent, that means resource is deleted and needs to be deleted from the Host as well

We'll be hard deleting transformed checks and certain checks older than N days, so this will prevent the host from bloating

**Comment by adityathebe** [comment]
> @adityathebe are there tests that cover this? We might want to make the page size configurable as well

We don't have test for reconciliation. Will create one

File: api/upstream.go [hunk 3]
```diff
@@ -8,14 +8,24 @@ import (
...
+
+type PaginateRequest struct {
+	Table string    `query:"table"`
+	From  uuid.UUID `query:"from"`
+	Size  int       `query:"size"`
+}
+
+type PaginateResponse struct {
+	Hash  string    `gorm:"column:sha256sum"`
+	Next  uuid.UUID `gorm:"column:last_id"`
```
Comment (yashmehrotra) [inline api/upstream.go:27]: Any reason for giving struct field and gorm column different names ?

Comment (adityathebe) [reply api/upstream.go:27]: I was just trying to avoid aliasing to `hash` as it's a SQL keyword.

**Comment by adityathebe** [comment]
PR is ready for review although my latest commit isn't showing up in here.

https://github.com/flanksource/incident-commander/commit/af139af5ec26453df9c47f33841bbe5b07aad0c5

**Pull request merged**